components/camel-zipfile/src/main/java/org/apache/camel/processor/aggregate/zipfile/ZipAggregationStrategy.java

```diff
@@ -178,81 +191,59 @@ public Exchange aggregate(Exchange oldExchange, Exchange newExchange) {
-        File tmpZip = Files.createTempFile(source.getName(), null).toFile();
-        tmpZip.delete();
-        if (!source.renameTo(tmpZip)) {
-            throw new IOException("Could not make temp file (" + source.getName() + ")");
-        }
-        byte[] buffer = new byte[8192];
 
-        FileInputStream fis = new FileInputStream(tmpZip);
-        ZipInputStream zin = new ZipInputStream(fis);
-        ZipOutputStream out = new ZipOutputStream(new FileOutputStream(source));
-
-        try {
-            InputStream in = new FileInputStream(file);
-            out.putNextEntry(new ZipEntry(fileName == null ? file.getName() : fileName));
-            for (int read = in.read(buffer); read > -1; read = in.read(buffer)) {
-                out.write(buffer, 0, read);
+    private static void newZipFile(File zipFile) throws URISyntaxException, IOException {
+        if (zipFile.exists()) {
+            if (!zipFile.delete()) { // Delete, because ZipFileSystem needs to create file on its own (with correct END bytes in the file)
+                throw new IOException("Cannot delete file " + zipFile);
             }
-            out.closeEntry();
-            IOHelper.close(in);
+        }
+        Map<String, String> env = new HashMap<>();
+        env.put("create", Boolean.TRUE.toString());
 
-            for (ZipEntry ze = zin.getNextEntry(); ze != null; ze = zin.getNextEntry()) {
-                out.putNextEntry(ze);
-                for (int read = zin.read(buffer); read > -1; read = zin.read(buffer)) {
-                    out.write(buffer, 0, read);
-                }
-                out.closeEntry();
-            }
-        } finally {
-            IOHelper.close(fis, zin, out);
+        try (FileSystem ignored = FileSystems.newFileSystem(getZipURI(zipFile), env)) {
+            //noop, just open and close FileSystem to initialize correct headers in file
         }
-        tmpZip.delete();
     }
-    
-    private static void addEntryToZip(File source, String entryName, byte[] buffer, int length) throws IOException {
-        File tmpZip = Files.createTempFile(source.getName(), null).toFile();
-        tmpZip.delete();
-        if (!source.renameTo(tmpZip)) {
-            throw new IOException("Cannot create temp file: " + source.getName());
-        }
 
-        FileInputStream fis = new FileInputStream(tmpZip);
-        ZipInputStream zin = new ZipInputStream(fis);
-        ZipOutputStream out = new ZipOutputStream(new FileOutputStream(source));
-        try {
-            out.putNextEntry(new ZipEntry(entryName));
-            out.write(buffer, 0, length);
-            out.closeEntry();
+    private void addFileToZip(File zipFile, File file, String fileName) throws IOException, URISyntaxException {
+        String entryName = fileName == null ? file.getName() : fileName;
+        Map<String, String> env = new HashMap<>();
+        env.put("useTempFile", Boolean.toString(this.useTempFile));
```

Comment:
So this parameter is something the JDK FileSystems understands and supports?

Reply:
This parameter delegated to underlying FileSystem. With **jar** URI is associated ZipFileSystem and it supports "useTempFile" property. It is not documented, but it is implemented since 1.7. Looking into source code http://www.docjar.com/html/api/com/sun/nio/zipfs/ZipFileSystem.java.html I have a bug there (This single property needs to be Boolean type, not String). I will fix this right now

Reply:
Fixed in commit 9afa571